staging/src/k8s.io/endpointslice/topologies/manager.go

```diff
@@ -0,0 +1,240 @@
+// various topologies.
+type Manager struct {
+	// topologiesByName maintains a mapping of a topology name to a Topology
+	// instance.
+	topologiesByName map[string]Topology
+	// defaultTopologyName is the name of the Topology which will be considered
+	// the default.
+	defaultTopologyName string
+
+	mu                        sync.Mutex
+	activeTopologiesByService map[string]map[discoveryv1.AddressType]Topology
+}
+
+// NewManager returns an instance of a topology Manager.
+//
+// `topologies` contains the set of topologies that this Mangaer will use and
+// manage. They should have unique names (as returned by their Topology.Names()
+// function). In the case when two topologies have the same name, the first
+// topology will be used.
+func NewManager(logger klog.Logger, defaultTopologyName string, topologies ...Topology) *Manager {
+	topologiesByName := make(map[string]Topology)
+	for _, topology := range topologies {
+		// Check if a topology with the same name already exists. In case of any
+		// conflict, we use the first occurrence.
+		_, ok := topologiesByName[topology.Name()]
+		if ok {
+			logger.Info("Found multilple topologies with the same name, using its first occurrence", "topologyName", topology.Name())
+			continue
+		}
+
+		topologiesByName[topology.Name()] = topology
+	}
+
+	return &Manager{
+		topologiesByName:          topologiesByName,
+		defaultTopologyName:       defaultTopologyName,
+		activeTopologiesByService: make(map[string]map[discoveryv1.AddressType]Topology),
+	}
+}
+
+// PopulateHints populates topology hints on EndpointSlices and returns updated
+// lists of EndpointSlices to create and update.
+//
+// It also updates all topology caches to reflect the latest topology in use.
+// This means that if a service was previously using topology A and is now using
+// topology B, the cache for topology A will be updated to remove any cached
+// hints for that service.
+func (m *Manager) PopulateHints(logger klog.Logger, service *corev1.Service, si *SliceInfo) ([]*discoveryv1.EndpointSlice, []*discoveryv1.EndpointSlice, []*EventBuilder) {
+	activeTopology, activeTopologyExists := m.activeTopologyForService(si.ServiceKey, si.AddressType)
+	desiredTopology, desiredTopologyEnabled, desiredTopologyExists := m.desiredTopologyForService(service)
+
+	// Identify if there is a change in the topology. Change means that the active
+	// topology is not equal to the desired topology.
+	if activeTopologyExists && (!desiredTopologyExists || activeTopology.Name() != desiredTopology.Name()) {
+		activeTopology.ClearCachedHints(logger, si.ServiceKey, si.AddressType)
+	}
+
+	if !desiredTopologyExists {
+		var events []*EventBuilder
+
+		if desiredTopologyEnabled {
+			// This is the case when the topology annotation has a non-disabled value
+			// but the topology does not exist in the topology manager. Send an event
+			// to notify about an unrecognizable topology.
```

Comment:
One of my goals with the recent KEP changes was to make "Hints" an optional implementation detail for topology heuristics, but not to necessarily require that every topology heuristic was implemented with hints. That would theoretically enable dataplane-only implementations of topology aware routing in the future for example. It _may_ make sense to leave some kind of carve out here to allow the EndpointSlice controller to be configured with allow-listed modes it would not warn for. We could also have some kind of pattern that we didn't warn for such as domain-prefixed strings, but since we haven't defined a pattern like this in the KEP, I think that would be too much of a stretch to add at this point.

Reply:
I think that this is still the goal, hints is just one mechanism to implement topology, but not the only one

Reply:
This comment is also related https://github.com/kubernetes/kubernetes/pull/121060/files#r1367545942, what do you think if we rename Topology by HintsHeuristics (or similar) to make it clear that this only works on the Hints fields